core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java

```diff
@@ -188,6 +204,47 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public static MultiSearchResponse fromXContext(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private static MultiSearchResponse.Item itemFromXContent(XContentParser parser) throws IOException {
+        // This parsing logic is complex here, because the multi search response is tricky.
+        // 1) The json objects inside the responses array are either a search response or a serialized exception
+        // 2) Each response json object gets a status field injected that ElasticsearchException.failureFromXContent(...) does not parse,
+        //    but SearchResponse.innerFromXContent(...) parses and then ignores. The status field is not needed to parse
+        //    the response item.
+
+        Token token = parser.nextToken();
+        assert token == Token.FIELD_NAME;
+
+        Item item = null;
+
+        String fieldName = null;
+        outer: for (; token != Token.END_OBJECT; token = parser.nextToken()) {
+            switch (token) {
+                case FIELD_NAME:
+                    fieldName = parser.currentName();
+                    if ("error".equals(fieldName)) {
+                        item = new Item(null, ElasticsearchException.failureFromXContent(parser));
+                    } else if ("status".equals(fieldName) == false) {
+                        item = new Item(SearchResponse.innerFromXContent(parser), null);
+                        break outer;
+                    }
+                    break;
+                case VALUE_NUMBER:
+                    if ("status".equals(fieldName)) {
+                        // Ignore the status value
+                    } else {
+                        throw new IllegalArgumentException("unexpected field name [" + fieldName + "]");
```

Comment:
I think it could be more permissive and just ignore the field? We used to be permissive when parsing responses in order to have a better forward compatibility...

Reply:
right, I've changed it.

Reply:
it should be permissive....but we are not testing forward comp properly because of the structure of these objects. This is quite a bummer, we should probably look into testing this, otherwise permissive or not doesn't make a difference as we don't test it.

Reply:
@javanna Don't we usually inject random fields to test this?

Reply:
yes but we discussed as part of this review issues around testing this, because the response contains an array of objects including exceptions. Exceptions still can hold metadata fields, hence injecting random fields in there would cause issues and needs to be skipped. It is complicated to distinguish between exceptions and search responses in the array, hence we don't currently insert random fields.

Reply:
Right, thanks @javanna, I've forgot this discussion. As long as it is permissive, even not tested, it's OK to be merged.

Reply:
Cool, I'll merge once the pr build is green.